### DISCUSSION TITLE
Help Inputting Data into SNN

### DISCUSSION THREAD

**Discussion by DS-567** [BindsNET/bindsnet#541] [General]
Hi,

I am using BindsNet for my research project to implement and train an SNN. In order to get familiar with BindsNet I am attempting to solve the XOR problem by creating a SNN, inputting the training data and store the network parameters. 

The XOR problem we are working on is for a 2 input gate. Therefore the data that we wish to use for training the network are the following combinations:

**(00) (01) (10) (11)** 

We want to use rate coding to convert a logic level 0 and 1 to a fixed frequency e.g. 
0 -> 10Hz spike train 
1 -> 20Hz spike train 

How do we feed the data into the SNN that we have created? Are there any examples showing how to do this with non-image based data?

**Comment by Hananel-Hazan** [comment]
The pixels in images in the MNIST examples are converted to spike train in Hz. All you need is the integer number to do the same, no image is necessary. You can use the function in the encoding library to create the signal base on an integer input and the function create the spike train in Hz, for example, see [Poisson doc](https://bindsnet-docs.readthedocs.io/bindsnet.encoding.html?highlight=Poisson#bindsnet.encoding.poisson)

**Reply by DS-567** [reply]
Ok I will have a look at that. Also I have another question. I cannot find anything that shows how to input your data and not data from the MNIST dataset. Is there any examples or documentation showing this?

**Reply by Hananel-Hazan** [reply]
There are other examples in the BindsNET directory that are not using MNIST, for example breakout, dotTracing or even the benchmark script.

**Reply by DS-567** [reply]
Hi again, thanks for your help. As I am using binary values of 0 and 1 as my input to the Poisson encoder, will it work feeding an integer 0 and 1 into the encoder? I seem to be getting all zeros in my encoded output. Do you know of any causes for this?

**Reply by Hananel-Hazan** [reply]
If the input for the SNN is already in binary (1 or 0), there is no need to use encoder. The Encoder method is to convert from frequency in Hz to spike train. Each encoder have its prefaces input range, Bernoulli for example can get real number between 0 to 1 and Poisson get non negative number. 
Please see the documentation for each one of the encoding, or suggest your own.

**Reply by DS-567** [reply]
Hi again Hananel, 

We are using BindsNet for supervised learning task with our datasets. We have encoded our spike trains in tensor format but do we need to assign a label to each sample of input data? And what is a label? Does label have to be different for every input or is label referring to the output neurons?

**Reply by Hananel-Hazan** [reply]
It depends on the task you are doing. The Unsupervised process means that during the training the network doesn't get any information about the label of the incoming spikes. However, you need to test the network somehow, so during the testing phase when the weights of the network don't change, you can test the accuracy of the network by comparing the output neurons to be discriminative enough and match the labels you have. Similar to what was done in the eth_minist example.

**Reply by DS-567** [reply]
Hannel, another question if you have time to answer. I have tried adding a hidden layer to my network to build the following topology: 

2 input neurons -> 4 hidden LIF neurons -> 2 output LIF neurons.

The input spike trains have not changed, all that has changed is the adding of this hidden layer.

Input data looks like the following:

# 0,0
zero_zero = {"Label": 0, "Image": torch.Tensor([[1, 1], [0, 0], [0, 0], [0, 0], [1, 1], [0, 0], [0, 0], [0, 0]]).byte()}

# 1,0
zero_one = {"Label": 1, "Image": torch.Tensor([[1, 1], [0, 0], [0, 1], [0, 0], [1, 1], [0, 0], [0, 1], [0, 0]]).byte()}

# 0,1
one_zero = {"Label": 1, "Image": torch.Tensor([[1, 1], [0, 0], [1, 0], [0, 0], [1, 1], [0, 0], [1, 0], [0, 0]]).byte()}

# 1,1
one_one = {"Label": 0, "Image": torch.Tensor([[1, 1], [0, 0], [1, 1], [0, 0], [1, 1], [0, 0], [1, 1], [0, 0]]).byte()}


For some reason now I am getting the error message: "RuntimeError: shape '[1, 4]' is invalid for input of size 8".

Do I need to use a different connection function to connects layers together that have different number of neurons or what do you think may be the issue?

**Reply by Hananel-Hazan** [reply]
I think that your running time is 8, and then the input nee to be in size of 2x8

There is two option to feed input to a neuron, spike or voltage. Feeding spikes can be done by feeding the input with the same time dimension of the network run time. You can see any of the mnist examples on how to convert a value of a pixel to a spike train and feed that to a neuron. On the other hand, feeding vmem needs to be done on the network level with the injection [method](https://github.com/BindsNET/bindsnet/blob/b5934eed8f7ecde39d96538ea01177c2db589c79/bindsnet/network/network.py#L271). Both methods need to be a matrix in size of the number of neurons and the running time of the network.
